core/shared/src/main/scala/zio/internal/FiberContext.scala

```diff
@@ -654,31 +659,29 @@ private[zio] final class FiberContext[E, A](
     }
 
   @tailrec
-  private[this] final def enterAsync(): Boolean = {
+  private[this] final def enterAsync(epoch: Long): Boolean = {
     val oldState = state.get
 
     oldState match {
       case Executing(_, observers, interrupt) =>
-        val newState = Executing(FiberStatus.Suspended(interruptible), observers, interrupt)
+        val interruptible = this.interruptible
+        val newState      = Executing(FiberStatus.Suspended(interruptible, epoch), observers, interrupt)
 
-        if (!state.compareAndSet(oldState, newState)) enterAsync()
-        else if (shouldInterrupt) {
-          // Fiber interrupted, so go back into running state:
-          exitAsync()
-          false
-        } else true
+        if (interruptible && interrupt) false
```

Comment:
@darl
I have a feeling this optimization does not make sense, i.e. we can't make decisions on _partial_ observed state until after we're out of the spinlock – that is, after we're in control and have the _current final_ version of the state. (that is, I'd move this if after the `if (!set)` as before)

Also, the tests on CircleCI keep failing – since `master` is building fine for some time now, I think the problem may be with this PR, could you please look into that?

Reply:
https://circleci.com/gh/zio/zio/35648 

```
[info]   - Stream.bufferSliding
[info]     + buffer the Stream with Error
[info]     - fast producer progress independently
[info]       Fiber failed.
[info]       A checked error was not handled.
[info]       None
[info]       
[info]       Fiber:145 was supposed to continue to:
[info]         a future continuation at zio.ZIO$$anonfun$run$2.apply(ZIO.scala:1120)
[info]         a future continuation at zio.ZIO$$anonfun$bracket_$1.apply(ZIO.scala:144)
[info]         a future continuation at zio.ZIO$$anonfun$run$2.apply(ZIO.scala:1120)
[info]         a future continuation at zio.ZIO$$anonfun$bracket_$1.apply(ZIO.scala:144)
[info]         a future continuation at zio.ZIO$$anonfun$run$2.apply(ZIO.scala:1120)
[info]         a future continuation at zio.ZManaged$$anonfun$use$1.apply(ZManaged.scala:649)
[info]         a future continuation at zio.test.Spec$$anonfun$foreachExec$1$$anonfun$apply$21.apply(Spec.scala:158)
[info]         a future continuation at zio.ZIOFunctions$$anonfun$foreachParN$2$$anonfun$apply$137$$anonfun$apply$139$$anonfun$apply$140$$anonfun$apply$141$$anonfun$apply$144.apply(ZIO.scala:2076)
[info]         a future continuation at zio.ZIO$$anonfun$forever$1.apply(ZIO.scala:454)
[info]       
[info]       Fiber:145 execution trace:
[info]         at zio.test.TestAspect$$anon$14$$anonfun$perTest$12$$anonfun$apply$19.apply(TestAspect.scala:427)
[info]         at zio.ZIO$$anonfun$raceWith$1$$anonfun$apply$79$$anonfun$apply$80$$anonfun$apply$92.apply(ZIO.scala:948)
[info]         at zio.internal.FiberContext$InterruptExit$$anonfun$apply$1.apply(FiberContext.scala:160)
[info]         at zio.internal.FiberContext$InterruptExit$.apply(FiberContext.scala:151)
[info]         at zio.ZIO$$anonfun$raceWith$1$$anonfun$apply$79$$anonfun$apply$80$$anonfun$apply$81$$anonfun$apply$82$$anonfun$apply$83$$anonfun$apply$85$$anonfun$apply$88$$anonfun$apply$91.apply(ZIO.scala:965)
```

Reply:
reverted it